controller/replica_controller.go

```diff
@@ -67,7 +67,8 @@ type ReplicaController struct {
 
 	instanceHandler *InstanceHandler
 
-	rebuildingLock *sync.Mutex
+	rebuildingLock          *sync.Mutex
+	inProgressRebuildingMap map[string]struct{}
```

Comment:
NIT: you can use map[string]interface as well then just push nil. If you only care about the key and not the value.
Both approaches are fine :)